### PR TITLE
Jira api integration

### PR DESCRIPTION
**Pull request by khushbu-maurya**

### CODE REVIEW

File: app/jobs/channelJobs.js [hunk 1]
```diff
@@ -528,16 +528,236 @@ var job = {
...
+
+            if (report) {
+                return {
+                    report: report,
+                    SchedulerLogsMeta: SchedulerLogsMeta
+                }
+
+            }
+        } catch (error) {
+
```
Comment (johibkhan2) [inline app/jobs/channelJobs.js]: log the exception

File: app/jobs/channelJobs.js [hunk 2]
```diff
@@ -528,16 +528,236 @@ var job = {
...
+        }
+    },
     createjiraTicket: async function (request) {
```
Comment (johibkhan2) [inline app/jobs/channelJobs.js]: this function is quite long...create a seprate function for setting props

File: app/jobs/channelJobs.js [hunk 3]
```diff
@@ -528,16 +528,236 @@ var job = {
...
+                            id: reportsData.SchedulerLogsMeta.SchedulerTaskLogId
+                        }
+                    }, { transaction });
+
+                await transaction.commit();
+
+
+                logger.log({
+                    level: 'info',
+                    message: 'jira issue is save into database'
```
Comment (johibkhan2) [inline app/jobs/channelJobs.js]: it should be "jira ticket is saved"

File: app/jobs/channelJobs.js [hunk 4]
```diff
@@ -528,16 +528,236 @@ var job = {
...
+                    jiraTicketLink: jiraSettings.record.organization + "/browse/" + jiraDetails.data.key,
+                });
+
+            }
+            catch (ex) {
+                transaction.rollback();
+
+                logger.log({
+                    level: 'error',
+                    message: 'error while saving jira issue into database',
```
Comment (johibkhan2) [inline app/jobs/channelJobs.js]: message should be "error occured while saving jira ticket"

File: app/jobs/channelJobs.js [hunk 5]
```diff
@@ -528,16 +528,236 @@ var job = {
...
+                    error: ex,
+                });
+                return { success: 0, message: ex };
+            }
+        }
+        else {
+            logger.log({
+                level: 'error',
+                message: 'error while creating jira',
```
Comment (johibkhan2) [inline app/jobs/channelJobs.js]: error occured while creating jira ticket

File: app/jobs/channelJobs.js [hunk 6]
```diff
@@ -528,16 +528,236 @@ var job = {
...
+            var issueList = [], jiraId = [];
+            var jiraSettings = await this.getJiraConfig();
+            var JiraTickets = await this.getCreatedTicketList();
+
+            JiraTickets.forEach(element => {
+                jiraId.push(element.jiraKey);
+            });
+
+            var jiraEndPoint = jiraSettings.record.organization + '/rest/api/3/search?jql=project in (' + jiraSettings.record.key + ') AND id in (' + jiraId.toString() + ') ';
+            if (request.status == "All") {
```
Comment (johibkhan2) [inline app/jobs/channelJobs.js]: remove muliple if else..use the command design pattern

File: app/jobs/channelJobs.js [hunk 7]
```diff
@@ -528,16 +528,236 @@ var job = {
...
+                    username: jiraSettings.record.userName,
+                    password: jiraSettings.record.apiToken,
+                }
+            })
+                .then((res) => {
+
+                    res.data.issues.forEach(element => {
+                        var issue = {};
+                        issue.issueID = element.id;
+                        issue.projectKey = element.fields.project.name;
```
Comment (johibkhan2) [inline app/jobs/channelJobs.js]: use a separate function to process response

File: app/jobs/channelJobs.js [hunk 8]
```diff
@@ -528,16 +528,236 @@ var job = {
...
+                        issue.summary = element.fields.summary;
+
+                        issueList.push(issue);
+                    });
+
+                })
+                .catch((error) => {
+                    logger.log({
+                        level: 'error',
+                        message: 'jira issue created with id : ' + res.id,
```
Comment (johibkhan2) [inline app/jobs/channelJobs.js]: message should be "error occured creating jira ticket"

File: app/jobs/channelJobs.js [hunk 9]
```diff
@@ -528,16 +528,236 @@ var job = {
...
+                return JiraTickets
+            }
+            else {
+                return { success: 0, message: "Jira Tickets is not found" };
+            }
+        }
+        catch (ex) {
+            logger.log({
+                level: 'error',
+                message: 'error while fetching Jira Tickets',
```
Comment (johibkhan2) [inline app/jobs/channelJobs.js]: error occured while fetching Jira Tickets

File: app/jobs/channelJobs.js [hunk 10]
```diff
@@ -528,16 +528,236 @@ var job = {
...
+                }
+            });
+            if (SchedulerLogs) {
+                return SchedulerLogs.jiraLink;
+            }
+            else{
+                return "Jira not created";;
+            }
+        } catch (error) {
+            return "Error while getting jira details";
```
Comment (johibkhan2) [inline app/jobs/channelJobs.js]: log the error and correct the message use error.getMessage() with log

File: app/jobs/jira-ticket.js [hunk 11]
```diff
@@ -0,0 +1,52 @@
+module.exports = {
+    "fields": {
```
Comment (johibkhan2) [inline app/jobs/jira-ticket.js]: why u need this?

File: app/jobs/schedulerJobs.js [hunk 12]
```diff
@@ -591,6 +587,53 @@ var job = {
                 reject({ message: 'error while generating image' + ex });
```
Comment (johibkhan2) [inline app/jobs/schedulerJobs.js]: correct the message

File: app/jobs/schedulerJobs.js [hunk 13]
```diff
@@ -591,6 +587,53 @@ var job = {
...
+    getSchedulerMetaData: async function (id) {
+        try {
+            var SchedulerLogs = await models.SchedulerTaskMeta.findOne({
+                where: {
+                    id: id
+                }
+            });
+            return SchedulerLogs;
+        } catch (error) {
+            return "Scheduler metadata is not found";
```
Comment (johibkhan2) [inline app/jobs/schedulerJobs.js]: log the error..use error.getMessage() in log

File: app/jobs/schedulerJobs.js [hunk 14]
```diff
@@ -591,6 +587,53 @@ var job = {
...
+                    log.viewTicket = await channelJobs.getJiraLink(logItem.SchedulerTaskMetum.id);
+                }
+                log.dashboardName = report.dashboard_name;
+                log.viewName = report.view_name;
+
+                log.descripition = report.mail_body;
+                log.isTicketCreated = logItem.isTicketCreated;
+                log.enableTicketCreation = logItem.enableTicketCreation;
+                log.comment = "";
+                log.task_executed = moment(logItem.task_executed).format("DD-MM-YYYY HH:mm")
```
Comment (johibkhan2) [inline app/jobs/schedulerJobs.js]: use this as constant DD-MM-YYYY HH:mm

File: app/report/reportService.js [hunk 15]
```diff
@@ -204,7 +204,7 @@ async function getScheduleReportLog(request) {
         throw { error: { message: `Error loading schedule report log ${e.message}` } };
```
Comment (johibkhan2) [inline app/report/reportService.js]: log the error also

File: app/send-team-notification.js [hunk 16]
```diff
@@ -132,7 +147,7 @@ exports.sendTeamNotification = async function sendNotification(teamConfig, repor
                         transaction.rollback();
                         logger.log({
                             level: 'error',
-                            message: 'error while sending team' + thresholdAlertEmail ? ' for threshold alert' : '',
+                            message: 'error while sending team' + reportData.report_obj.thresholdAlert ? ' for threshold alert' : '',
```
Comment (johibkhan2) [inline app/send-team-notification.js]: error occurred

File: app/jobs/jira-ticket.js (hunk 11 above)
Comment (khushbu-maurya) [reply app/jobs/jira-ticket.js]: this file is proto type of create jira ticket useing this file json in create api

File: app/jobs/channelJobs.js [hunk 17]
```diff
@@ -528,16 +532,180 @@ var job = {
...
+            if (report) {
+                return {
+                    report: report,
+                    SchedulerLogsMeta: SchedulerLogsMeta
+                }
+
+            }
+        } catch (error) {
+            return {
+                message: "Scheduler logs metadata not found "
```
Comment (johibkhan2) [inline app/jobs/channelJobs.js]: message should be "Scheduler logs metadata is not found"

**Pull request merged**